crates/rattler_repodata_gateway/src/fetch/jlap/mod.rs

```diff
@@ -0,0 +1,800 @@
+//!     }
+//! }
+//! ```
+//!
+//! ## TODO
+//!
+//! The following items still need to be implemented before this module should be considered
+//! complete:
+//!  - Use the checksum to validate our JLAP file after we update it
+//!  - Our tests do not exhaustively cover our error states. Some of these are pretty easy to
+//!    trigger (e.g. invalid JLAP file or invalid JSON within), so we should definitely make
+//!    tests for them.
+use crate::fetch::cache::Blake2b256;
+
+use blake2::Digest;
+use reqwest::{
+    header::{HeaderMap, HeaderValue},
+    Client, Response,
+};
+use serde::{Deserialize, Serialize};
+use std::path::{Path, PathBuf};
+use std::str;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use url::Url;
+
+/// File suffix for JLAP file
+pub const JLAP_FILE_SUFFIX: &str = "jlap";
+
+/// File name of JLAP file
+pub const JLAP_FILE_NAME: &str = "repodata.jlap";
+
+/// File suffix for JLAP files
+pub const JLAP_FOOTER_OFFSET: usize = 2;
+
+/// Represents the variety of errors that we come across while processing JLAP files
+#[derive(Debug, thiserror::Error)]
+pub enum JLAPError {
+    #[error(transparent)]
+    /// Pass-thru for JSON errors found while parsing JLAP file
+    JSONParseError(serde_json::Error),
+
+    #[error(transparent)]
+    /// Pass-thru for JSON errors found while patching
+    JSONPatchError(json_patch::PatchError),
+
+    #[error(transparent)]
+    /// Pass-thru for HTTP errors encountered while requesting JLAP
+    HTTPError(reqwest::Error),
+
+    #[error(transparent)]
+    /// Pass-thru for file system errors encountered while requesting JLAP
+    FileSystemError(tokio::io::Error),
+
+    #[error("No patches found in the JLAP file")]
+    /// Error returned when JLAP file has no patches in it
+    NoPatchesFoundError,
+
+    #[error("No matching hashes can be found in the JLAP file")]
+    /// Error returned when none of the patches match the hash of our current `repodata.json`
+    NoHashFoundError,
+
+    #[error("Hash from the JLAP metadata and hash from updated repodata file do not match")]
+    /// Error when we have mismatched hash values after updating our `repodata.json` file
+    /// At this point, we should consider the `repodata.json` file corrupted and fetch a new
+    /// version from the server.
+    HashesNotMatchingError,
+}
+
+/// Represents the numerous patches found in a JLAP file which makes up a majority
+/// of the file
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Patch {
+    /// Next hash of `repodata.json` file
+    pub to: String,
+
+    /// Previous hash of `repodata.json` file
+    pub from: String,
+
+    /// Patches to apply to current `repodata.json` file
+    pub patch: json_patch::Patch, // [] is a valid, empty patch
+}
+
+/// Represents the metadata for a JLAP file, which is typically found at the very end
+#[derive(Serialize, Deserialize, Debug)]
+pub struct JLAPMetadata {
+    /// URL of the `repodata.json` file
+    pub url: String,
```

Comment:
You can use the `Url` type here.

Reply:
I tried this but these aren't actually URLs 🙃. It's just the filename (e.g. `repodata.json` or `current_repodata.json`) 🤷‍♂️

Reply:
Will do 👍